src/testinferred.jl

```diff
@@ -0,0 +1,96 @@
+module TestInferred
+
+export @testinferred
+
+using InteractiveUtils: gen_call_with_extracted_types
+using Test: @test
+
+"""
+    @testinferred [AllowedType] f(x)
+
+Tests that the call expression `f(x)` returns a value of the same type inferred by the compiler.
+It is useful to check for type stability.
+This is similar to `Test.@inferred`, but instead of throwing an error, a `@test` is added.
+
+Optionally, `AllowedType` relaxes the test, by making it pass when either the type of `f(x)` matches the inferred type modulo `AllowedType`, or when the return type is a subtype of `AllowedType`.
+This is useful when testing type stability of functions returning a small union such as `Union{Nothing, T}` or `Union{Missing, T}`.
+"""
+macro testinferred(ex)
+    return _inferred(ex, __module__)
+end
+
+macro testinferred(allow, ex)
+    return _inferred(ex, __module__, allow)
+end
+
+# helper functions
+_args_and_call((args..., f)...; kwargs...) = (args, kwargs, f(args...; kwargs...))
+_materialize_broadcasted(f, args...) = Broadcast.materialize(Broadcast.broadcasted(f, args...))
+@static if isdefined(Base, :typesplit)
+    const typesplit = Base.typesplit
+else
+    Base.@pure function typesplit(@nospecialize(a), @nospecialize(b))
+        a <: b && return Base.Bottom
+        isa(a, Union) && return Union{typesplit(a.a, b), typesplit(a.b, b)}
+        return a
+    end
+end
```

Comment:
Is this block starting with `@static if isdefined` different from the one in ConstInferred. Can we share the typesplit? Or better yet, maybe we just want to restrict to newer Julia versions where this is anyway available.

(I have to admit I even forgot what this function does).